File: examples/python/heap_1.py
def heap_insert(heap, x):
    heap.append(x)
    i = len(heap) - 1
    while i > 0:
        p = (i - 1) // 2
        if heap[p] <= heap[i]:
            break
        heap[p], heap[i] = heap[i], heap[p]
        i = p


def heap_extract(heap):
    if not heap:
        return None
    top = heap[0]
    heap[0] = heap.pop()
    i, n = 0, len(heap)
    while True:
        smallest = i
        l, r = 2 * i + 1, 2 * i + 2
        if l < n and heap[l] < heap[smallest]:
            smallest = l
        if r < n and heap[r] < heap[smallest]:
            smallest = r
        if smallest == i:
            break
        heap[i], heap[smallest] = heap[smallest], heap[i]
        i = smallest
    return top


if __name__ == "__main__":
    h = []
    heap_insert(h, 5)
    heap_insert(h, 2)
    heap_insert(h, 8)
    print(heap_extract(h))
    print(heap_extract(h))

File: examples/ruby/heap_1.rb
def heap_insert(heap, x)
  heap << x
  i = heap.size - 1
  while i > 0
    p = (i - 1) / 2
    break if heap[p] <= heap[i]
    heap[p], heap[i] = heap[i], heap[p]
    i = p
  end
end

def heap_extract(heap)
  return nil if heap.empty?
  top = heap[0]
  heap[0] = heap.pop
  i, n = 0, heap.size
  loop do
    smallest = i
    l, r = 2 * i + 1, 2 * i + 2
    smallest = l if l < n && heap[l] < heap[smallest]
    smallest = r if r < n && heap[r] < heap[smallest]
    break if smallest == i
    heap[i], heap[smallest] = heap[smallest], heap[i]
    i = smallest
  end
  top
end

h = []
heap_insert(h, 5)
heap_insert(h, 2)
heap_insert(h, 8)
puts heap_extract(h)
puts heap_extract(h)
